src/main/java/com/google/sps/servlets/utils/Statistics.java

```diff
@@ -14,29 +14,89 @@
 
 package com.google.sps.servlets.utils;
 
-import com.google.sps.servlets.Range;
+import java.math.BigDecimal;
+import java.util.Comparator;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
 public class Statistics {
-  // Aggregated score for a list of comments
+  // Contains sentiment scores in the range [-1, 1] with given intervals.
   private Map<Range, Integer> aggregateValues;
   private double averageScore;
+  private static final double INTERVAL = 0.2;
+  private static final double UPPER_END = 1.0;
+  private static final double LOWER_END = -1.0;
+
+  public Map<Range, Integer> getAggregateValues() {
+    return aggregateValues;
+  }
+
+  public double getAverageScore() {
+    return averageScore;
+  }
 
   public Statistics(List<Double> sentimentScores) {
     setAggregateScores(sentimentScores);
     setAverageScore(sentimentScores);
   }
 
+  /**
+   * Categorize all score values into different range intervals and count the frequency for each
+   * interval, and set the aggregatedValues.
+   *
+   * @param sentimentScores a list of score values from -1.0 to 1.0
+   */
   private void setAggregateScores(List<Double> sentimentScores) {
-    // TODO(Xin): Add sorting code
     aggregateValues = new HashMap<>();
-    aggregateValues.put(new Range(-1.0, 1), 1);
+    BigDecimal curPoint = BigDecimal.valueOf(LOWER_END);
+    BigDecimal interval = BigDecimal.valueOf(INTERVAL);
+
+    // Initialize the HashMap with intervals
+    while (curPoint.doubleValue() < UPPER_END) {
+      Range currentRange =
+          new Range(
+              curPoint.doubleValue(), Math.min(curPoint.add(interval).doubleValue(), UPPER_END));
+      curPoint = curPoint.add(interval);
+      aggregateValues.put(currentRange, 0);
+    }
+
+    // Add score's interval to different ranges two sorting with and two pointers pop-up
+    sentimentScores.sort(Comparator.naturalOrder());
+    curPoint = BigDecimal.valueOf(LOWER_END);
+    int scoreIdx = 0;
+    while ((scoreIdx < sentimentScores.size()) && (curPoint.doubleValue() < UPPER_END)) {
+      double scoreVal = sentimentScores.get(scoreIdx);
+      // return null for edge cases
+      if ((scoreVal < LOWER_END) || (scoreVal > UPPER_END)) {
+        scoreIdx += 1;
+        continue;
+      }
+      // check available interval for scoreVal
+      BigDecimal nextPoint = curPoint.add(interval);
```

Comment:
I'm not sure if there's any benefit to using `BigDecimal` if all of our comparisons use `double` anyway (other than perhaps our intervals accumulating less floating point imprecision).

Which test cases were failing before you switching to using `BigDecimal`?

Reply:
Discussed in the chat about logic behind BigDecimal.